File: src/leadforge/scrapers/angi.py
import structlog

from leadforge.scrapers.base import BaseAPIClient

logger = structlog.get_logger()


class AngiClient(BaseAPIClient):
    """Angi (Angie's List) business scraper."""

    def __init__(self):
        super().__init__(base_url="https://www.angi.com", timeout=30.0)

    async def search_business(self, name: str, zip_code: str) -> dict | None:
        """Search for a business on Angi. Returns business data or None."""
        try:
            client = await self._get_client()
            response = await client.get(
                "/search",
                params={"query": name, "zipCode": zip_code},
                headers={"Accept": "text/html"},
            )
            response.raise_for_status()
            text = response.text

            import json
            import re

            script_match = re.search(
                r'<script[^>]*type="application/ld\+json"[^>]*>(.*?)</script>',
                text,
                re.DOTALL,
            )
            if script_match:
                try:
                    ld_json = json.loads(script_match.group(1))
                    return {
                        "angi_name": ld_json.get("name"),
                        "angi_rating": ld_json.get("aggregateRating", {}).get(
                            "ratingValue"
                        ),
                        "angi_review_count": ld_json.get("aggregateRating", {}).get(
                            "reviewCount"
                        ),
                    }
                except json.JSONDecodeError:
                    pass

            return None
        except Exception as e:
            logger.warning("angi_search_failed", error=str(e), name=name)
            return None

File: api/src/scrapers/angi.ts
// =py scrapers/angi
import { aggregateRating, encodeParams, extractLdJson, fetchText } from './base';

export interface AngiProfile {
  angi_name: string | null;
  angi_rating: number | null;
  angi_review_count: number | null;
}

// =py search_business — informational only; the enrichment caller logs the result and stores nothing
export async function searchBusiness(name: string, zipCode: string): Promise<AngiProfile | null> {
  try {
    const params = encodeParams({ query: name, zipCode });
    const html = await fetchText(`https://www.angi.com/search?${params}`, { headers: { Accept: 'text/html' } });

    const ldJson = extractLdJson(html);
    if (!ldJson) return null;

    const rating = aggregateRating(ldJson);
    return {
      angi_name: (ldJson.name as string) ?? null,
      angi_rating: (rating.ratingValue as number) ?? null,
      angi_review_count: (rating.reviewCount as number) ?? null,
    };
  } catch (error) {
    console.warn('angi_search_failed', { name, error: error instanceof Error ? error.message : String(error) });
    return null;
  }
}
